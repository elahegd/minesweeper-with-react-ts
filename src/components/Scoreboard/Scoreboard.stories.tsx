import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Scoreboard, ScoreboardProps } from './Scoreboard';

export default {
    title: "Scoreboard/Scoreboard",
    component: Scoreboard
} as Meta;

const Template: Story<ScoreboardProps> = (args) => <Scoreboard {...args} />;

export const ScoreboardExample = Template.bind({});
ScoreboardExample.args = {
    timer: "000",
    levels: ['beginner', 'intermediat', 'expert'],
    defaultValue: 'beginner',
    bombs: "010",
}