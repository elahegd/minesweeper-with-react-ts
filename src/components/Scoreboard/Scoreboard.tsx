import React, { FC, ChangeEvent } from 'react';
import styled from "@emotion/styled";

import { Reset } from "./Reset";
import { Counter } from "./Counter";
import { Level } from './Level';

export interface ScoreboardProps {
    timer: string;
    levels: string[];
    defaultValue?: string;
    onChangeLevel: (event: ChangeEvent<HTMLSelectElement>) => void;
    onReset: () => void;
    bombs: string;
}

export const Scoreboard: FC<ScoreboardProps> = ({timer, levels, defaultValue, onChangeLevel, onReset, bombs}) => {
  return (
    <Wrapper>
      <Counter>{timer}</Counter>
      <ResetAndLevelWrapper>
        <Level value={defaultValue} onChange={onChangeLevel}>
          {levels}
        </Level>
        <Reset onReset={onReset} />
      </ResetAndLevelWrapper>
      <Counter>{bombs}</Counter>
    </Wrapper>
  )
}

const Wrapper = styled.div`
  display: flex;
  width: 100%;
  justify-content: space-between;
`;
const ResetAndLevelWrapper = styled.div`
  display: flex;
`;