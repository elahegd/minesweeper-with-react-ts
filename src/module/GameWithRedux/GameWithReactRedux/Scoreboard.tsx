import React, { FC, useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { RootState } from "../../../pages/store";
import { actions } from "../index";

import { Scoreboard as ScoreboardComponent } from "../../../components/Scoreboard";
import { GameLevel, LevelNames } from '../../GameSettings';

export const Scoreboard: FC = () => {
    const dispatch = useDispatch();

    const { time, bombs, flagCounter, level } = useSelector(
        ({ game: { time, bombs, flagCounter, level } }: RootState) => ({ 
            time, bombs, flagCounter, level 
        })
    );

    const handleChange = useCallback(
        ({target: {value: level}}: React.ChangeEvent<HTMLSelectElement>) => {
          dispatch(actions.changeLevel(level as LevelNames))
          }, 
          // Stryker disable next-line ArrayDeclaration
          []);

    const onReset = useCallback(
        () => dispatch(actions.reset()), 
        // Stryker disable next-line ArrayDeclaration
        []
    );

    return (
        <>
            <ScoreboardComponent 
                timer={String(time)} 
                levels={GameLevel as unknown as string[]}
                defaultValue={level}
                onChangeLevel={handleChange}
                onReset={onReset}
                bombs={String(bombs - flagCounter)}
            />
        </>
    )
}
