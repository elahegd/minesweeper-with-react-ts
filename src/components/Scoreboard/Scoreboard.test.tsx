import React from 'react';
import { render, screen, } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import {Scoreboard} from "./Scoreboard";

describe("Render Scoreboard component with props", () => {
    it("Scoreboard renders correctly", () => {
        const { asFragment } = render(<Scoreboard
            levels= {['beginner', 'intermediat', 'expert']}
            timer= {"000"}
            onReset={() => null} 
            onChangeLevel={() => null}
            defaultValue={'intermediat'} bombs={"010"} />);

        expect(asFragment()).toMatchSnapshot();;
    });
});

describe("Change handler behaviour", () => {
    it("Change from intermediate to expert", () => {
        const onChange = jest.fn();

        render(<Scoreboard onChangeLevel={onChange} 
            levels= {['beginner', 'intermediat', 'expert']}
            timer= {"000"}
            onReset={() => null} 
            defaultValue={'intermediat'} bombs={"010"} />);

        userEvent.selectOptions(screen.getByRole('combobox'), 'expert');

        expect(screen.getByRole('option', {name : 'expert'})).toBeEnabled();
        expect(onChange).toHaveBeenCalled();
    })
})