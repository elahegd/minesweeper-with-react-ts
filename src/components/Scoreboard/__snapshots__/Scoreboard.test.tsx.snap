// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Scoreboard component with props Scoreboard renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1n7y7g1"
  >
    <div
      class="css-d2q1pa"
    >
      000
    </div>
    <div
      class="css-ov1ktg"
    >
      <select
        class="css-1lpmfg0"
      >
        <option
          class="css-zcmdxr"
          value="beginner"
        >
          beginner
        </option>
        <option
          class="css-zcmdxr"
          value="intermediat"
        >
          intermediat
        </option>
        <option
          class="css-zcmdxr"
          value="expert"
        >
          expert
        </option>
      </select>
      <div>
        <button
          class="css-1kb3z2h"
        >
          ðŸ˜€
        </button>
      </div>
    </div>
    <div
      class="css-d2q1pa"
    >
      010
    </div>
  </div>
</DocumentFragment>
`;
